Add CI runner for `pull_request`.

Grinkers opened the pull request:
It would be nice to have Github show CI runners right in the PRs.

I removed the `paths` bit because I don't think it's needed?

w-henderson commented:
I think we do need the `paths` bit - it ensures we only run the tests we need to based on what's changed...?

w-henderson commented:
That being said, if I've messed up the dependencies please feel free to change them, but I don't think we want to remove them.

Grinkers commented:
> That being said, if I've messed up the dependencies please feel free to change them, but I don't think we want to remove them.

I don't think anything was wrong, I just sort of assumed we should run everything. For example a change in `humphrey-server/` might break something in `examples/`. I tend to just run everything all the time. This crate compiles so fast that I don't think we need to worry about that. Having two sets of paths would also be super ugly.

That being said, I'll leave the choice up to you.  At the end of the day, just wanted to see CI in PR threads directly.

w-henderson commented:
Okay, that's a fair point - if you could just put the `paths` back for the JSON spec tests (as they really definitely can't be affected by anything else), then this looks good. Can you also make sure that it only runs on pushes to master then, as PRs cover pushes to branches.

Grinkers commented:
Rebased/conflicts resolved. No other changes yet.

Grinkers commented:
> Okay, that's a fair point - if you could just put the `paths` back for the JSON spec tests (as they really definitely can't be affected by anything else), then this looks good. Can you also make sure that it only runs on pushes to master then, as PRs cover pushes to branches.

I could, but I guess I wonder what the optimization here is? It seems pretty harmless to run everything all the time (and keeps the yml simple). It just takes a few seconds to run. It's also quite nice that I get CI reports when working on a non-master branch (but can also see how this could be undesirable).

Just an alternate perspective. I'll let you have last say.

w-henderson commented:
I'd rather not have everything running twice in PRs. I think my philosophy is that when you're not on master, you don't expect the CI to pass, everything is a work-in-progress. However, once you're at the stage where you want to merge your changes to master with a PR, that's when you're interested in the CI reports.

I don't think we always need the JSON spec tests running, it's really irrelevant unless you're working directly on the JSON code, in which case they will run.

Grinkers commented:
Now using the old json spec paths.

Added filters for `master` and `main`, so if a rename of the default branch happens, it's not breaking. 
Added colors.

w-henderson reviewed (approved):
Looks good, thank you!
